lightning-invoice/src/utils.rs

```diff
@@ -229,67 +232,85 @@ where
 		// phantom route into the invoice, and let the sender find the path to the `real_node_pubkey`
 		// node by looking at our public channels.
-		if route_hints.is_empty() {
-			route_hints.push(RouteHint(vec![]))
-		}
-		for route_hint in &mut route_hints {
-			route_hint.0.push(RouteHintHop {
-				src_node_id: real_node_pubkey,
-				short_channel_id: phantom_scid,
-				fees: RoutingFees {
-					base_msat: 0,
-					proportional_millionths: 0,
-				},
-				cltv_expiry_delta: MIN_CLTV_EXPIRY_DELTA,
-				htlc_minimum_msat: None,
-				htlc_maximum_msat: None,});
-		}
+		let empty_route_hints = route_hints.len() == 0;
+		let mut have_pushed_empty = false;
+		let route_hints = route_hints
+			.chain(core::iter::from_fn(move || {
+				if empty_route_hints && !have_pushed_empty {
+					// set flag of having handled the empty route_hints and ensure empty vector
+					// returned only once
+					have_pushed_empty = true;
+					Some(RouteHint(Vec::new()))
+				} else {
+					None
+				}
+			}))
+			.map(move |mut hint| {
+				hint.0.push(RouteHintHop {
+					src_node_id: real_node_pubkey,
+					short_channel_id: phantom_scid,
+					fees: RoutingFees {
+						base_msat: 0,
+						proportional_millionths: 0,
+					},
+					cltv_expiry_delta: MIN_CLTV_EXPIRY_DELTA,
+					htlc_minimum_msat: None,
+					htlc_maximum_msat: None,
+				});
+				hint
+			});
 
 		phantom_hints.push(route_hints);
+
 	}
 
 	// We have one vector per real node involved in creating the phantom invoice. To distribute
 	// the hints across our real nodes we add one hint from each in turn until no node has any hints
 	// left (if one node has more hints than any other, these will accumulate at the end of the
 	// vector).
-	let mut invoice_hints: Vec<RouteHint> = Vec::new();
-	let mut hint_idx = 0;
+	rotate_nested_vectors(phantom_hints)
+}
 
-	loop {
-		let mut remaining_hints = false;
+/// Draw items iteratively from multiple nested vectors.  The items are retrieved by index and
+/// rotates through the vectors - first the zero index then the first index then second index, etc.
+fn rotate_nested_vectors<T, I: Iterator<Item = T>>(mut vecs: Vec<I>) -> impl Iterator<Item = T> {
+	let mut idx = 0;
 
-		for hints in phantom_hints.iter() {
-			if invoice_hints.len() == 3 {
-				return invoice_hints
+	core::iter::from_fn(move || {
+		let mut exhausted_vectors = 0;
+		loop {
+			if vecs.is_empty() {
+				return None;
 			}
-
-			if hint_idx < hints.len() {
-				invoice_hints.push(hints[hint_idx].clone());
-				remaining_hints = true
+			let next_idx = idx % vecs.len();
+			let hint_opt = vecs[next_idx].next();
+			idx += 1;
+			if let Some(hint) = hint_opt {
+				return Some(hint);
+			}
+			// exhausted_vectors increase when the "next_idx" vector is exhausted
+			exhausted_vectors += 1;
+			// return None when all of the nested vectors are exhausted
+			if exhausted_vectors > vecs.len() {
+				return None;
 			}
```

Comment:
Won't this overcount a vector that is exhausted early? For example, what if the input vector contains one iterator that yields no items and one iterator with many items? Not enough items will be yielded by the returned iterator.

Reply:
I thought the same thing when I first worked through Matt's solution.  It took me a bit of time but I think the key realization is that when the call site calls `next` the counter is set to zero. The counter will only be equal to the len of `vecs` when the loop has hit each of the nested vectors with a `None` response (all exhausted).

For example in this test: https://github.com/lightningdevkit/rust-lightning/blob/d4d8b81b85487c79e6787a7d542949e00de49144/lightning-invoice/src/utils.rs#L2004 in the iteration after `b3` is returned, the counter will accumulate to 2 and then return `c1`. The iteration after that the loop goes through each of the nested vectors which are exhausted and then the counter equals the len of `vecs`

Reply:
Ah, yes. I see. Thanks for explaining.